Operators article (#76)
TCCPP/wiki: wiki/cpp-tutorial/operators.md

> # Operators

Reviewer: Operator precedence and associativity are very important concepts here and aren't really covered. It's ok to defer to another PR (or for someone else to write), but if so please leave `wip: true`

> In any programming language, the core standard provides simple ways to do various operations on numbers, variables, pointers, even functions with operator overloading. In C++, we are also provided with various operators that we can use to do different tasks. Usually they have the same symbols and methodology as elementary mathematics. We will discuss what are operators and how are they separated into groups by the C++ standard below.

Reviewer: Tone-wise, this is a bit heavy / textbook-like. We're trying to keep these relatively to the point (as done on other pages).

|`/`|Division|`10 / 2`|`5`|
|`%`|Modulus/Reminder|`10 % 3`|`1`|
|`++`|Increment|`int a = 5; a++`|`6`|
|`--`|Decrement|`int a = 5; a--`|`4`|

### Basic Examples for Arithmetic Operators:

```c++
int a = 10;
int b = 3;

int sum = a + b;        // 13
int difference = a - b; // 7
int product = a * b;    // 30
int quotient = a / b;   // 3 (integer division)
int remainder = a % b;  // 1
```

### Division Behavior

Division in C++ behaves differently based on the types of the values or variables involved. Some examples are given below:

```c++
int x = 7 / 2;      // 3 (integer division, decimal part truncated)
double y = 7.0 / 2; // 3.5 (floating-point division)
double z = 7 / 2.0; // 3.5 (at least one operand is floating-point)
```

### Increment and Decrement

The increment (`++`) and decrement (`--`) operators have two forms: prefix and postfix.

```c++
int a = 5;
int b = ++a; // Prefix: increment a first, then assign. a = 6, b = 6
int c = a++; // Postfix: assign a first, then increment. a = 7, c = 6

int x = 10;
int y = --x; // Prefix: decrement x first, then assign. x = 9, y = 9
int z = x--; // Postfix: assign x first, then decrement. x = 8, z = 9
```

## Assignment Operators

> Assignment operators are used to assign values to variables. The basic assignment operator is `=`, but C++ provides compound assignment operators that combine arithmetic operations with assignment. The assignment is done from left to right, the value of left is assigned the value on right, in some cases after an operation.

Reviewer: ```suggestion
Assignment operators are used to assign values to variables. The basic assignment operator is `=`, but C++ provides compound assignment operators that combine arithmetic operations with assignment. The assignment is done from right to left, the value of the right hand side is assigned the left hand side, in some cases after an operation.
```

|Operator|Description|Example|Result|
|---|---|---|---|
|`==`|Equal to|`5 == 5`|`true`|
|`!=`|Not equal to|`5 != 3`|`true`|
|`>`|Greater than|`5 > 3`|`true`|
|`<`|Less than|`5 < 3`|`false`|
|`>=`|Greater than or equal to|`5 >= 5`|`true`|
|`<=`|Less than or equal to|`5 <= 3`|`false`|

### Examples

```c++
int a = 10;
int b = 20;

bool is_equal = (a == b);        // false
bool is_not_equal = (a != b);    // true
bool is_greater = (a > b);       // false
bool is_less = (a < b);          // true
bool is_greater_equal = (a >= 10); // true
bool is_less_equal = (b <= 20);  // true
```

These operators are commonly used in conditional statements:

```c++
int age = 18;

if (age >= 18) {
    std::cout << "You are an adult.\n";
}
```

> [!WARNING]
> 
> Be careful not to confuse `=` (assignment) with `==` (comparison). Using `=` in a conditional statement is a common mistake that can lead to unexpected behavior.

## Logical Operators

Logical operators are used to combine multiple boolean expressions or to invert boolean values. They are essential for creating complex conditional statements.

|Operator|Description|Example|Result|
|---|---|---|---|
|`&&`|Logical AND|`true && false`|`false`|

> |`\|`|Logical OR|`true \| false`|`true`|

Reviewer: ```suggestion
|`\|\|`|Logical OR|`true \| false`|`true`|
```

```c++
const unsigned int READ = 0b100;
const unsigned int WRITE = 0b010;
unsigned int permissions = READ | WRITE; // 0b110
```

### Bitwise XOR (`^`)

Sets each bit to 1 if the corresponding bits are different. If they are the same, set them to 0.

```c++
unsigned int a = 12; // 1100 in binary
unsigned int b = 10; // 1010 in binary
unsigned int result = a ^ b; // 0110 in binary = 6
```

Common use: toggling bits or simple encryption.

```c++
int value = 42;
int key = 123;
int encrypted = value ^ key;
int decrypted = encrypted ^ key; // Back to 42
```

### Bitwise NOT (`~`)

Inverts all bits (1 becomes 0, 0 becomes 1).

```c++
unsigned int a = 12; // 00001100 in binary (assuming 8 bits)
unsigned int result = ~a; // 11110011 in binary
```

### Left Shift (`<<`)

Shifts bits to the left, filling with zeros. Each shift left multiplies the value by 2.

```c++
unsigned int a = 5;  // 0101 in binary
unsigned int result = a << 2; // 10100 in binary = 20
```

### Right Shift (`>>`)

Shifts bits to the right. Each shift right divides the value by 2 (integer division).

```c++
unsigned int a = 20; // 10100 in binary
unsigned int result = a >> 2; // 00101 in binary = 5
```

> [!NOTE]
> 

> > For signed integers, right shift behavior is implementation-defined. Some systems perform arithmetic shift (sign extension), while others perform logical shift (zero fill).

Reviewer 2: There is nothing implementation-defined here. Right-shift for signed integers is an arithmetic right shift. https://eel.is/c++draft/expr.shift#3

> [!NOTE] For Advanced Programmers
> 
> Be aware of type promotion when using bitwise operators. Small integer types like `char` and `short` are promoted to `int` before bitwise operations, which can lead to unexpected results with the NOT operator. Also, shifting by an amount greater than or equal to the bit width of the type results in undefined behavior.

## Other Operators

> C++ provides several other operators that don't fit into the previous groups but are essential for various programming tasks. Do keep in mind this operators are useful as you learn more about C++ so feel free to use them whenever you get the chance.

Reviewer: ```suggestion
C++ provides several other operators that don't fit into the previous groups but are essential for various programming tasks. Do keep in mind these operators are useful as you learn more about C++ so feel free to use them whenever you get the chance.
```

### Ternary Operator (`? :`)

> The ternary operator is a shorthand for `if-else` statements. It takes three operands.

Reviewer 2: It may be better to use the formal name "conditional operator". https://eel.is/c++draft/expr.cond

"Ternary operator" is colloquial.

**Syntax:** `condition ? value_if_true : value_if_false`

```c++
int a = 10;
int b = 20;
int max = (a > b) ? a : b; // max = 20

> // This is Equivalent to:

Reviewer: ```suggestion
// This is equivalent to:
```

int max;
if (a > b) {
    max = a;
} else {
    max = b;
}
```

### Comma Operator (`,`)

The comma operator evaluates multiple expressions from left to right and returns the value of the rightmost expression.

```c++
int a = (5, 10, 15); // a = 15

> int x = 1, y = 2;

Reviewer: In the context of teaching the comma operator, using a comma in declarations might be confusing. We also in general recommend against using commas in declarations like this.

int result = (++x, ++y); // x = 2, y = 3, result = 3
```

> It's most commonly seen in `for` loops:

Reviewer: We probably want to hedge this whole section as "this is a clever trick that is situationally useful but should almost always be avoided"

```c++

> for (int i = 0, j = 10; i < j; ++i, --j) {

Reviewer: Ditto here

    // Multiple operations in the increment section
}
```

### Sizeof Operator

The `sizeof` operator returns the size (in bytes) of a type or variable.

```c++
int size_of_int = sizeof(int);        // Usually 4
int size_of_double = sizeof(double);  // Usually 8

int array[10];
int array_size = sizeof(array);       // 40 (assuming int is 4 bytes)

> int array_length = sizeof(array) / sizeof(array[0]); // 10

Reviewer: Let's maybe omit the array example. It's a dated and error-prone example, not something we want to showcase to people who are learning.

```

### Member Access Operators

These operators are used to access members of structures, classes, and pointers.

```c++
struct Point {
    int x;
    int y;
};

Point p;
p.x = 10;     // Dot operator for direct access
p.y = 20;

Point* ptr = &p;
ptr->x = 30;  // Arrow operator for pointer access
ptr->y = 40;

// ptr->x is equivalent to (*ptr).x
```

### Address-Of and Dereference Operators

The address-of operator (`&`) returns the memory address of a variable, while the dereference operator (`*`) accesses the value at a given address.

```c++
int value = 42;
int* ptr = &value;  // ptr holds the address of value
int result = *ptr;  // result = 42 (dereference ptr)

*ptr = 100;         // Changes value to 100
```
> [!NOTE]
> 

> > `&` cannot be used with a null value, meanwhile a pointer can be null. You would learn more on pointers later in the wiki.

Reviewer: What is a "null value"? This should be rewritten or removed

The scope resolution operator is used to access global variables, static members, and namespace members.

```c++
int value = 10; // Global variable

int main() {
    int value = 20; // Local variable
    std::cout << value << '\n';   // 20 (local)
    std::cout << ::value << '\n'; // 10 (global)
}

// Accessing namespace members
std::cout << "Hello\n";

// Accessing static class members
class MyClass {
public:
    static int count;
};
int x = MyClass::count;
```

### Cast Operators

C++ provides several cast operators for type conversion (C++11 and later also includes more sophisticated casting mechanisms).

```c++
double pi = 3.14159;
int truncated = (int)pi;           // C-style cast
int truncated2 = static_cast<int>(pi); // C++ style cast (preferred)
```

> [!NOTE] For Advanced Programmers
> 
> C++ provides four named cast operators: `static_cast`, `dynamic_cast`, `const_cast`, and `reinterpret_cast`. These provide more type safety and clarity than C-style casts and should be preferred in modern C++. Additionally, operator precedence and associativity become important when combining multiple operators in complex expressions.

## Experimenting with Operators

> As you continue learning C++, you'll discover that operators can be used in creative and powerful ways. Some of this areas are operator overloading, combining operators, bitwise tricks, etc.

Reviewer: ```suggestion
As you continue learning C++, you'll discover that operators can be used in creative and powerful ways. Some of these areas are operator overloading, combining operators, bitwise tricks, etc.
```

The best way to master operators is through practice. Try:

- Writing programs that use different combinations of operators
- Experimenting with operator precedence to understand evaluation order
- Implementing algorithms that leverage bitwise operations
- Creating your own classes with overloaded operators

> Remember that readable code is often more important than clever tricks. Use operators in ways that make your intent clear to other programmers (or yourself in the future).

Reviewer: Per the writing guide "don't be preachy or lecture-like" / "don't be dismissive of common mistakes":

```suggestion
We highly recommend writing code that is clear and readable to others (and your future self) and avoid clever tricks that can make code harder to read.
```

> [!NOTE] For Advanced Programmers
> 

> > Consider exploring topics like expression templates, SFINAE with operators, the spaceship operator (`<=>`) introduced in C++20, and how compilers optimize operator usage. The `constexpr` keyword can also be applied to operator overloads for compile-time evaluation. Understanding move semantics and perfect forwarding becomes crucial when overloading assignment operators and implementing efficient operator overloads.

Reviewer: I think this is too broad and generic, maybe just omit it
